### DISCUSSION TITLE
JsonEncodingException when passing a new model instance to a sub-component

### DISCUSSION THREAD

**Discussion by leecolarelli** [livewire/livewire#3035] [7. Bugs (archived)]
### Description
I'm getting a JsonEncodingException with the following code... but only when the $gigApplication is passed to a laravel subcomponent (none livewire, tried both anonymous and normal component). Its a simple component as it only needs to receive the variable and display some buttons based on data attributes. So I did not think it necessary to use a livewire component here.

```
public ?GigApplication $gigApplication;
public function storeGigApplication($boost = null)
    {
        $this->gigApplication = new GigApplication;
        $this->gigApplication->gigr()->associate($this->gigr);
        $this->gigApplication->gig()->associate($this->gig);
        $this->message = 'Gig application received!'; // Show an application received message
        if ($boost == 'boost') {
            $this->gigApplication->boosted = 1;
            $this->message = 'Gig application received and boosted!'; // Show an application received and boosted message
        }
        $this->gigApplication->save();

        $this->boostAvailable = false; // Set boostAvailable to false
    }
```

The property state has to be optional as its null to begin with, but then its a model instance when one is created via the method. The instance stores to the database correctly, and if i manually refresh the page both components work as expected. In otherwords, the variable is passed to the subcomponent correctly.

Also, if I remove the subcomponent or don't pass the variable, it also works fine. So the issue is definitely with the variable getting passed somewhere in the refresh.

I'm not competent enough yet to debug the issue, hence posting here.

its passed to subcomponent like this `<x-gigr.gig-list.gig-list-item-buttons :gigApplication="$gigApplication" :boostAvailable="$boostAvailable"/>`

The model itself only contains two foreign keys and 6 boolean attributes, so there's nothing to trip up json. 

**Context**
- Livewire version: 2.4.1
- Laravel version: 8.35.1
- Browser: Chrome on windows 10, latest stable release.

**Comment by leecolarelli** [comment]
Forgot to mention, if I find() the model instance in tinker and convert `->toJson()` it converts without errors.

**Comment by leecolarelli** [comment]
The only way I have gotten round this at the moment is to have a number of property states in my parent component, and pass these to the sub-component rather than passing the model instance itself. It works but it's a far messier solution.

**Comment by joshhanley** [comment]
@leecolarelli can you also share the code for your blade component?

**Comment by leecolarelli** [comment]
> @leecolarelli can you also share the code for your blade component?

Hi Josh, sure. I also saw your reply on Discord :) I'll just keep it here for clarity.

gig-list-item.blade.php...
```
{{-- Gigr Gig List Item Component todo: get transition working --}}
<div x-data
     x-transition:enter-start="opacity-0 transition ease-out duration-1000"
     x-transition:enter-end="opacity-100 transition ease-in duration-1000">

    {{-- Flash Message --}}
    <div>
        @if ($message)
            <div class="flex items-center bg-green-500 text-white text-sm font-bold px-4 py-3 mb-4" role="alert">
                <svg class="fill-current w-4 h-4 mr-2" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20">
                    <path
                        d="M12.432 0c1.34 0 2.01.912 2.01 1.957 0 1.305-1.164 2.512-2.679 2.512-1.269 0-2.009-.75-1.974-1.99C9.789 1.436 10.67 0 12.432 0zM8.309 20c-1.058 0-1.833-.652-1.093-3.524l1.214-5.092c.211-.814.246-1.141 0-1.141-.317 0-1.689.562-2.502 1.117l-.528-.88c2.572-2.186 5.531-3.467 6.801-3.467 1.057 0 1.233 1.273.705 3.23l-1.391 5.352c-.246.945-.141 1.271.106 1.271.317 0 1.357-.392 2.379-1.207l.6.814C12.098 19.02 9.365 20 8.309 20z"/>
                </svg>
                <p>{{ $message }}</p>
            </div>
        @endif
    </div>

    <div class="flex">

        {{-- Gig Info --}}
        <div class="flex-initial">

            {{-- Gigr Gig List Item Status Component todo: change back to passing just the $gigApplication when works --}}
            <x-gigr.gig-list.gig-list-item-status :applied="$applied" :boosted="$boosted" :offered="$offered"
                                                  :rejected="$rejected" :accepted="$accepted" :declined="$declined"/>

            <p class="font-bold">£value (calculate from
                rate- {{ $gig->rate_value }}
                ratetype-
                {{ $gig->rate_type }})</p>
            {{-- if gig category has a parent --}}
            <p>
                @if ($gig->gigCategory->parent)
                    <span class="font-bold">
                                    {{ $gig->gigCategory->parent->name }}:
                                </span>
                @endif
                <span class="font-bold">{{ $gig->gigCategory->name }}</span>
            </p>
            <p class="">{{ $gig->start_date }} - {{ $gig->end_date }}
                {{ $gig->time_of_day }}</p>
            <p class="mb-1">rating required - {{ $gig->rating_required }}</p>
            <p class="mb-1">experience {{ $gig->experience_required }}</p>
            <p class="mb-1">payment type - {{ $gig->payment_type }} payment timescale -
                {{ $gig->payment_timescale }} (is this not included?)
            </p>
            <p class="mb-1">gig type - {{ $gig->gig_type }} (is this not included?)</p>
            <p class="mb-1">hours - {{ $gig->hours_available }} (is this not
                included?)</p>
            <p class="mb-1">address - {{ $gig->address->address }}</p>
            {{-- Show the address distance if the gig has one --}}
            @if ($gig->address->distance)
                <p class="mb-1">distance - {{ $gig->address->distance }}</p>
            @endif
        </div>

        {{-- Gig Rating and Buttons --}}
        <div class="relative flex-1">
            <div class="absolute top-0 right-0">
                <p>bizr rating (23)</p>
            </div>

            {{-- Gigr Gig List Item Buttons Component todo: change back to passing just the $gigApplication and $boostAvailable when works--}}
            <x-gigr.gig-list.gig-list-item-buttons :applied="$applied" :boosted="$boosted" :offered="$offered"
                                                   :accepted="$accepted" :declined="$declined"
                                                   :boostAvailable="$boostAvailable"/>

        </div>
    </div>

    {{-- if admin --}}
    <div class="bg-gray-100 p-5 mt-3">
        <h5><u>Admin</u></h5>
        <p class="mb-1">ID: {{ $gig->id }}</p>
        <p class="mb-1">Status: {{ $gig->status }}</p>
        <p class="mb-1">Created At: {{ $gig->created_at }}</p>
        <p class="mb-0">Updated At: {{ $gig->updated_at }}</p>
    </div>

</div>
```

**Comment by leecolarelli** [comment]
As you can see, passing all the $variables to the sub-components is extremely messy. I could pass as an array instead but I'd rather just be able to pass the model instance successfully!

**Comment by joshhanley** [comment]
@leecolarelli yep, easier to keep here.

Sorry I wasn't clear, I meant the sub component blade code that you think is causing issues 🙂  the `gig-list-item-buttons` right?

Is the blade code you shared above the Livewire component blade code?

I'm struggling a bit currently to see how it all works, and sometimes it can be something small in an unexpected location causing the issue.

If you can outline how it all fits together and share all the pieces I need to see how it would run that would be great, so we can try and get to the bottom of it.

Or even better, if you can share a sample repo that demonstrates the issue that I could pull down and test, that would be a big help.

Thanks!

**Comment by leecolarelli** [comment]
Josh, here is gig-list-item-buttons.blade.php...

```
{{-- Gigr Gig List Item Buttons Component --}}

<div class="absolute bottom-0 right-0 text-right">

    {{-- If the gigr has NOT applied for the gig --}}
    @if(!$applied)

        <button wire:click="storeGigApplication()"
                class="uppercase font-bold font-gigr-blue block disabled:opacity-50">
            Apply
        </button>
        {{-- And if no one has boosted an application --}}
        @if ($boostAvailable)
            <button wire:click="storeGigApplication('boost')"
                    class="uppercase font-bold font-gigr-blue block disabled:opacity-50">Apply & Boost
            </button>
        @endif

        {{-- If the gigr HAS applied for the gig --}}
    @else

        {{-- If no one has boosted an application --}}
        @if (!$boosted && $boostAvailable)
            <button wire:click="boostGigApplication()"
                    class="uppercase font-bold font-gigr-blue block disabled:opacity-50">Boost
            </button>
        @elseif(!$boosted && !$boostAvailable)
            <p class="uppercase font-bold font-gigr-blue block">Boost Unavailable</p>
        @endif

    @endif

    {{-- If the gig application has been offered the gig --}}
    @if($offered)

        {{-- If the gig application has NOT been accepted/declined --}}
        @if(!$accepted && !$declined)
            <button wire:click="acceptGig()"
                    class="uppercase font-bold font-gigr-blue block disabled:opacity-50">
                Accept Gig
            </button>
            <button wire:click="declineGig()"
                    class="uppercase font-bold font-gigr-blue block disabled:opacity-50">
                Decline Gig
            </button>
        @endif

        {{-- If the gig application HAS been accepted and NOT marked as completed--}}
        @if($accepted && !$completed)
            {{-- Mark As Completed Button todo: needs popup confirmation --}}
            <button wire:click="completed()"
                    class="uppercase font-bold font-gigr-blue block disabled:opacity-50">
                <span>Mark As Completed</span>
            </button>
        @endif

    @endif

</div>
```

The Apply button on gig-list-item-buttons.blade.php calls storeGigApplication() on the parent component, which then creates and stores a GigApplication model instance. At this point the main (parent) component refreshes, and the `gig-list-item-buttons.blade.php` should no longer show the apply button etc. 

FYI the variables in gig-list-item-buttons.blade.php should be `$gigApplication->applied`, `$gigApplication->boosted` etc but they are not currently as I cannot pass the $gigApplication successfully.

To reiterate, the JsonEncodingException ONLY happens when i pass the $gigApplication to the subcomponent like this `<x-gigr.gig-list.gig-list-item-buttons :gigApplication="$gigApplication" :boostAvailable="$boostAvailable"/>`.

And even then, the exception is not there if i refresh the page... upon refresh everything works correctly, and the $gigApplication is passed to the subcomponent correctly.

I wondered if it was because before the GigApplication model instance is created, the $gigApplication being passed to the subcomponent is NULL, but upon creating the model instance it is now an object. However, as i said, I do not fully understand the refresh process so I do not know.

If this is all still too confusing then let me know and i'll try and find time to make an example repository.

Thank you.

**Comment by joshhanley** [comment]
@leecolarelli thanks! The blade file all looks ok. I pulled it down and tested it and it seemed to all work fine for me. I created a gig application model and it created one without issue and re-passed it into the component without the error you are getting. But I had to disabled a lot to get it to run as I don't have your full setup.

If you can share a repo that would be great 🙂 happy to DM on discord if you'd prefer to keep it private.
